src/callbacks_step/save_restart.jl

```diff
@@ -156,5 +161,34 @@ function load_restart_file(semi::AbstractSemidiscretization, restart_file)
     load_restart_file(mesh_equations_solver_cache(semi)..., restart_file)
 end
 
+"""
+    load_adaptive_time_integrator!(integrator, restart_file::AbstractString)
+
+Load the context information for error-based time integrators saved in a `restart_file`.
+"""
+function load_adaptive_time_integrator!(integrator, restart_file::AbstractString)
+    controller = integrator.opts.controller
+    # Read context information for controller
+    h5open(restart_file, "r") do file
+        # Ensure that necessary information was saved
+        if !("time_integrator_qold" in keys(attributes(file))) ||
+           !("time_integrator_dtpropose" in keys(attributes(file))) ||
+           (hasproperty(controller, :err) &&
+            !("time_integrator_controller_err" in keys(attributes(file))))
+            error("Missing data in restart file: check the consistency of adaptive time controller with initial setup!")
+        end
+        integrator.qold = read(attributes(file)["time_integrator_qold"])
+        integrator.dtpropose = read(attributes(file)["time_integrator_dtpropose"])
+        # Accept step to use dtpropose already in the first step
+        integrator.accept_step = true
+        # Reevaluate integrator.fsal_first on the first step
+        integrator.reeval_fsal = true
+        # Load additional parameters for PIDController
+        if hasproperty(controller, :err)
+            controller.err[:] = read(attributes(file)["time_integrator_controller_err"])
+        end
```

Comment:
This has to be behind some additional checks which controller is used. We default to `PIDController`s in many cases, but other algorithms use a `PIController`. See https://docs.sciml.ai/DiffEqDocs/stable/extras/timestepping/

Reply:
I'm not sure, that I understood correctly. Do you mean to check before explicitly whether controller is `PIDController` or `PIController`? If yes, it requires a dependency on `OrdinaryDifffEq`, where these types are defined. This package is huge and Michael said better not to depend on it for efficiency.
So check `hasproperty(controller, :err)` is exactly to distinguish between PID and PI without loading OrdinaryDiffEq.
So I'm not sure what I need to change. Now it works both for `PIDController` and `PIController`

Reply:
Please explain this in a comment

Reply:
Okey, I tested with `SSPRK43` in [one of the tests](https://github.com/trixi-framework/Trixi.jl/pull/1565/files#diff-43aaf8164d76b0a0bff6056e14898a55083136429d17b91aad6d67bd85e2473b)